File: sgf-poker/sgf-poker-ios/Core/UseCases/CalculatePrizesUseCase.swift
// Core/UseCases/CalculatePrizesUseCase.swift

import Foundation

/// Calculates prize amounts and bounty pool using the league's fixed rules.
final class CalculatePrizesUseCase {

    func execute(game: Game, players: [Player]) -> PrizeCalculation {
        let presentGamePlayers = game.players.filter { $0.present }
        let presentCount = presentGamePlayers.count

        // Resolve full Player for each GamePlayer to check membership
        let presentPlayers: [(gp: GamePlayer, player: Player?)] = presentGamePlayers.map { gp in
            (gp, players.first { $0.id == gp.playerId })
        }

        // Entry fees: members pay 25, non-members pay 30
        let memberCount    = presentPlayers.filter { $0.player?.isMember == true || $0.player?.isFounder == true }.count
        let nonMemberCount = presentCount - memberCount
        let entryFeeTotal  = Double(memberCount) * PrizeRules.memberEntryFee
                           + Double(nonMemberCount) * PrizeRules.nonMemberEntryFee

        // Rebuys: $25 each
        let totalRebuys = presentGamePlayers.reduce(0) { $0 + $1.rebuyCount }
        let rebuyTotal  = Double(totalRebuys) * PrizeRules.rebuyAmount

        let totalPool  = entryFeeTotal + rebuyTotal
        let bountyPool = totalPool * PrizeRules.bountyShare
        let prizePool  = totalPool * PrizeRules.prizeShare

        // Prize distribution based on present player count
        let distribution = PrizeRules.distribution(for: presentCount)

        // Build prizes for ranked players only, sorted by position
        let ranked = presentGamePlayers
            .filter  { $0.hasFinished }
            .sorted  { ($0.finalPosition ?? Int.max) < ($1.finalPosition ?? Int.max) }

        let prizes: [PlayerPrize] = ranked.enumerated().compactMap { index, gp in
            guard index < distribution.count else { return nil }
            let player    = players.first { $0.id == gp.playerId }
            let name      = player?.name ?? "Unknown"
            let entryFee  = (player?.isMember == true || player?.isFounder == true)
                            ? PrizeRules.memberEntryFee
                            : PrizeRules.nonMemberEntryFee
            let amount    = prizePool * distribution[index]
            return PlayerPrize(
                id: gp.id,
                playerName: name,
                position: gp.finalPosition ?? (index + 1),
                entryFee: entryFee,
                prizeAmount: amount
            )
        }

        let unrankedCount = presentCount - ranked.count

        return PrizeCalculation(
            presentCount: presentCount,
            memberCount: memberCount,
            nonMemberCount: nonMemberCount,
            totalRebuys: totalRebuys,
            totalPool: totalPool,
            bountyPool: bountyPool,
            prizePool: prizePool,
            distribution: distribution,
            prizes: prizes,
            unrankedCount: unrankedCount
        )
    }
}

File: sgf-poker-android/app/src/main/java/com/sgf/poker/usecases/prizes/CalculatePrizesUseCase.java
package com.sgf.poker.usecases.prizes;

import com.sgf.poker.domain.model.*;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

/** Calculates prize amounts and bounty pool using fixed league rules. */
public class CalculatePrizesUseCase {

    public PrizeCalculation execute(Game game, List<Player> players) {
        var presentGPs = game.getPresentPlayers();
        int presentCount = presentGPs.size();

        int memberCount = (int) presentGPs.stream().filter(gp -> {
            var p = findPlayer(players, gp.getPlayerId());
            return p != null && p.isEligibleForMemberFee();
        }).count();
        int nonMemberCount = presentCount - memberCount;

        double totalCollected = memberCount    * PrizeRules.MEMBER_ENTRY_FEE
                             + nonMemberCount * PrizeRules.NON_MEMBER_ENTRY_FEE;
        double nonMemberFeeTotal = nonMemberCount * (PrizeRules.NON_MEMBER_ENTRY_FEE - PrizeRules.MEMBER_ENTRY_FEE);
        double entryFeeTotal = totalCollected - nonMemberFeeTotal;

        int totalRebuys = presentGPs.stream().mapToInt(GamePlayer::getRebuyCount).sum();
        double rebuyTotal = totalRebuys * PrizeRules.REBUY_AMOUNT;

        double totalPool  = entryFeeTotal + rebuyTotal;
        double bountyPool = totalPool * PrizeRules.BOUNTY_SHARE;
        double prizePool  = totalPool * PrizeRules.PRIZE_SHARE;

        var distribution = PrizeRules.distribution(presentCount);

        var ranked = presentGPs.stream()
                .filter(GamePlayer::hasFinished)
                .sorted((a, b) -> {
                    int pa = a.getFinalPosition() != null ? a.getFinalPosition() : Integer.MAX_VALUE;
                    int pb = b.getFinalPosition() != null ? b.getFinalPosition() : Integer.MAX_VALUE;
                    return Integer.compare(pa, pb);
                })
                .collect(Collectors.toList());

        var prizes = IntStream.range(0, Math.min(ranked.size(), distribution.size()))
                .mapToObj(i -> {
                    var gp = ranked.get(i);
                    var player = findPlayer(players, gp.getPlayerId());
                    String name = player != null ? player.getName() : "Unknown";
                    double fee  = player != null && player.isEligibleForMemberFee()
                            ? PrizeRules.MEMBER_ENTRY_FEE : PrizeRules.NON_MEMBER_ENTRY_FEE;
                    int pos = gp.getFinalPosition() != null ? gp.getFinalPosition() : i + 1;
                    return new PlayerPrize(gp.getId(), name, pos, fee, prizePool * distribution.get(i));
                })
                .collect(Collectors.toList());

        return new PrizeCalculation(
                presentCount, memberCount, nonMemberCount, totalRebuys,
                totalCollected, nonMemberFeeTotal, totalPool,  bountyPool, prizePool, distribution, prizes,
                presentCount - ranked.size());
    }

    private Player findPlayer(List<Player> players, String id) {
        return players.stream().filter(p -> p.getId().equals(id)).findFirst().orElse(null);
    }
}
